Add rpath for hipfft-test to help find libomp.

malcolmroberts opened the pull request.

malcolmroberts committed:
Add rpath for hipfft-test to help find libomp.

regan-amd commented on clients/tests/CMakeLists.txt:
```diff
@@ -98,6 +98,15 @@ endif()
+    target_link_libraries( hipfft-test PRIVATE OpenMP::OpenMP_CXX
+                           -L${HIP_CLANG_ROOT}/lib -Wl,-rpath=${HIP_CLANG_ROOT}/lib )
+  endif()
+endif()
```
I would have a few questions:
1. was there a reason to move this out of the `foreach` body? I'd expect the option `BUILD_CLIENTS_TESTS_OPENMP` to apply equally to all test targets, _e.g._, to `hipfft_mpi_worker` as well if desired;
2. some changes seem unrelated to me, _e.g._, the compilation option `-DBUILD_CLIENTS_TESTS_OPENMP` is dropped and the addition of an ad-hoc include path is omitted for `CMAKE_CXX_COMPILER` other than `*hipcc` or `*clang++`. It seems fine to me as the former is irrelevant to the current form of the target sources, and the latter cannot happen given [this requirement](https://github.com/ROCm/hipFFT/blob/3ee87897daf268eaeceba16028625624b4bce5f5/clients/CMakeLists.txt#L70), in my understanding. Could you confirm?

regan-amd commented on clients/tests/CMakeLists.txt:
```diff
@@ -98,6 +98,15 @@ endif()
+    message( STATUS "OpenMP is not supported on CUDA, building tests without it" )
+  else()
+    target_link_libraries( hipfft-test PRIVATE OpenMP::OpenMP_CXX
+                           -L${HIP_CLANG_ROOT}/lib -Wl,-rpath=${HIP_CLANG_ROOT}/lib )
```
Suggestion: would adding "`INSTALL_RPATH_USE_LINK_PATH TRUE`" to the `set_target_properties` within the body of the `foreach` below be an acceptable option instead? I feel like that would generalize better, including all possible targets and their dependencies.

evetsso replied:
Using the CMake properties is more robust than providing the command line options directly.  Plus, with the properties you also get to specify BUILD_RPATH and INSTALL_RPATH separately, which can matter in some situations.

regan-amd committed:
Define static class member of hipfft_params for hipfft_mpi_worker

TorreZuk commented:
rocBLAS and hipBLAS are now using this scheme https://github.com/ROCm/rocm-libraries/pull/624

malcolmroberts commented:
> rocBLAS and hipBLAS are now using this scheme [ROCm/rocm-libraries#624](https://github.com/ROCm/rocm-libraries/pull/624)

@TorreZuk : it looks to me like roc/hipBLAS won't be able to find all of the libraries that it needs if one installs to, for example, ~/rocBLAS.  Am I mistaken about this?  Has this been tested?

TorreZuk commented:
> > rocBLAS and hipBLAS are now using this scheme [ROCm/rocm-libraries#624](https://github.com/ROCm/rocm-libraries/pull/624)
> 
> @TorreZuk : it looks to me like roc/hipBLAS won't be able to find all of the libraries that it needs if one installs to, for example, ~/rocBLAS. Am I mistaken about this? Has this been tested?

Here is the hipBLAS version tested with gcc and clang https://github.com/ROCm/hipBLAS/pull/1050.  Tests reported along with normal CI and some manual tests (inc. windows).   Are you installing the clients to some unique location?   If so, there is runpath information still being set.   This change prefers the cmake config for openmp amd clang that was @estewart08  addition to rocm.

malcolmroberts commented:
> > > rocBLAS and hipBLAS are now using this scheme [ROCm/rocm-libraries#624](https://github.com/ROCm/rocm-libraries/pull/624)
> > 
> > 
> > @TorreZuk : it looks to me like roc/hipBLAS won't be able to find all of the libraries that it needs if one installs to, for example, ~/rocBLAS. Am I mistaken about this? Has this been tested?
> 
> Here is the hipBLAS version tested with gcc and clang [ROCm/hipBLAS#1050](https://github.com/ROCm/hipBLAS/pull/1050). Tests reported along with normal CI and some manual tests (inc. windows). Are you installing the clients to some unique location? If so, there is runpath information still being set. This change prefers the cmake config for openmp amd clang that was @estewart08 addition to rocm.

I'm looking at the use-case where the installation directory isn't /opt/rocm.  We have use cases where users will install rocFFT to their home directory, so, yes, the clients may be installed to some unique location.

estewart08 commented:
> > > > rocBLAS and hipBLAS are now using this scheme [ROCm/rocm-libraries#624](https://github.com/ROCm/rocm-libraries/pull/624)
> > > 
> > > 
> > > @TorreZuk : it looks to me like roc/hipBLAS won't be able to find all of the libraries that it needs if one installs to, for example, ~/rocBLAS. Am I mistaken about this? Has this been tested?
> > 
> > 
> > Here is the hipBLAS version tested with gcc and clang [ROCm/hipBLAS#1050](https://github.com/ROCm/hipBLAS/pull/1050). Tests reported along with normal CI and some manual tests (inc. windows). Are you installing the clients to some unique location? If so, there is runpath information still being set. This change prefers the cmake config for openmp amd clang that was @estewart08 addition to rocm.
> 
> I'm looking at the use-case where the installation directory isn't /opt/rocm. We have use cases where users will install rocFFT to their home directory, so, yes, the clients may be installed to some unique location.

ROCm strongly prefers relative paths for rpath from what I understand. If there are install locations outside of a ROCm directory structure then I would think LD_LIBRARY_PATH would need set.

malcolmroberts commented:
> > Here is the hipBLAS version tested with gcc and clang [ROCm/hipBLAS#1050](https://github.com/ROCm/hipBLAS/pull/1050). Tests reported along with normal CI and some manual tests (inc. windows). Are you installing the clients to some unique location? If so, there is runpath information still being set. This change prefers the cmake config for openmp amd clang that was @estewart08 addition to rocm.
> 
> I'm looking at the use-case where the installation directory isn't /opt/rocm. We have use cases where users will install rocFFT to their home directory, so, yes, the clients may be installed to some unique location.

Working with
```
set_target_properties( ${target} PROPERTIES BUILD_RPATH "${HIP_CLANG_ROOT}/lib" )
set_target_properties( ${target} PROPERTIES INSTALL_RPATH "$ORIGIN/../llvm/lib" )
```
and installing to a user-local directory, the installed version of hipff-test doesn't find libomp.so.

malcolmroberts committed:
Use INSTALL_RPATH_USE_LINK_PATH for linking test executables.

regan-amd reviewed (dismissed):
Thank you!

evetsso commented:
Rpath is being used for multiple purposes:
* Finding libhipfft.so
* Finding runtime libraries (libamdhip64.so)
* Finding libomp.so

If any of these are installed to different prefixes, then a single rpath entry won't be enough to find all of them.  Also, we don't want absolute install rpaths, as @estewart08 points out.  Build rpaths may be absolute.

So we can have `${HIP_CLANG_ROOT}/lib` as a build rpath, but install rpath will need to be something relative to `$ORIGIN`.  If different libraries are in different prefixes, then IMO it's reasonable to expect users to have to change their configuration to make them findable by the dynamic linker.  Rpath isn't really meant to solve this whole problem.  Users could either change /etc/ld.so.conf.d or set LD_LIBRARY_PATH.

malcolmroberts committed:
openmp_LIB_DIR style rpath for tests when OpenMP CONFIG works.
Fix linking directory to fix clients compilation.
Allow for building from clients/tests.

malcolmroberts commented:
Ok, in the current version, tests installed to non-rocm directories need LD_LIBRARY_PATH set for libomp.so to be found.  For the upcoming OpenMP configuration, we can use openmp_LIB_DIR to help with linking.

Also, hipfft.h is changed from including hipfft-export.h to including hipfft/hipfft-export.h, which fixes the test-only build.

malcolmroberts committed:
Make clients/tests find hipfft only of not in hipfft build scope

regan-amd commented on library/include/hipfft/hipfft.h:
```diff
@@ -45,8 +45,8 @@
 
-#include "hipfft-export.h"
-#include "hipfft-version.h"
+#include "hipfft/hipfft-export.h"
+#include "hipfft/hipfft-version.h"
```
I am wary of this change. In my understanding, once installed, `hipfft.h`, `hipfft-export.h`, and `hipfft-version.h` are all placed in the same directory, the last two files being automatically created when building/installing the hipfft library (they are not natively in the repository). So, for anything that includes an **installed library**'s `hipfft.h`, the compiler should typically find the required `hipfft-export.h`, and `hipfft-version.h` (without any change to this file) since the directory of that included `hipfft.h` is the first place it should explore when using such double quotes `#include` ([reference](https://en.cppreference.com/w/cpp/preprocessor/include.html#Notes)).

I'm actually suspecting that the test-only build failed because it was not including an installed library's `hipfft.h` (say `/opt/rocm/include/hipfft/hipfft.h`) but your local repository's header instead, due to
`target_include_directories( ${target}
    PRIVATE
    [...]
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../library/include>
  )`
being used (unconditionally of `HIPFFT_BUILD_SCOPE`) in `clients/tests/CMakeLists.txt`.
In fact, I do not think that we need to add that include directory explicitly: if hipfft is built/installed alongside the tests, cmake should use the built/installed library's header automatically; if doing a test-only build, cmake should find the installed library's header from the `target_link_libraries( ${target} PRIVATE hip::hipfft [...])` command (given how the public `target_include_directories` of hipfft itself are [defined](https://github.com/ROCm/hipFFT/blob/4df4ddfb77e71b98e853822e4ce9b92c71b8d079/library/CMakeLists.txt#L97)).

Could we try reverting this file's changes and removing `$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../library/include>` from the tests' `target_include_directories` instead?

regan-amd commented on clients/tests/hipfft_mpi_worker.cpp:
```diff
@@ -23,6 +23,9 @@
 #include "../../shared/mpi_worker.h"
 #include "../hipfft_params.h"
 
+// initialize static class member of hipfft_params
+std::vector<gpubuf> hipfft_params::externally_managed_workareas = std::vector<gpubuf>();
+
```
FYI: this may be removed as that was resolved in https://github.com/ROCm/hipFFT/pull/174

regan-amd commented on clients/tests/CMakeLists.txt:
```diff
@@ -98,6 +111,14 @@ endif()
 string( CONCAT TESTS_OUT_DIR "${PROJECT_BINARY_DIR}" ${TESTS_OUT_DIR} )
 
 option( BUILD_CLIENTS_TESTS_OPENMP "Build tests with OpenMP" ON )
+if( BUILD_CLIENTS_TESTS_OPENMP )
```
Suggestion: maybe we could add `AND BUILD_WITH_LIB NOT STREQUAL "CUDA"` above, to avoid a fatal build error if building with cuda (openMP is ignored thereafter in that case, anyways)?

regan-amd commented on clients/tests/CMakeLists.txt:
```diff
@@ -98,6 +111,14 @@ endif()
+  #find_package( OpenMP CONFIG PATHS "${HIP_CLANG_ROOT}/lib/cmake" )
+  if( NOT OPENMP_FOUND OR NOT DEFINED ${openmp_LIB_DIR} )
+     # Fall-back to module mode.
+     find_package( OpenMP REQUIRED )
```
Would this ever define `openmp_LIB_DIR`? If or when it does, is the `${HIP_CLANG_ROOT}/${openmp_LIB_DIR}` used below consistent with it regardless of _which_ `OpenMP` package is found?

regan-amd commented on clients/tests/CMakeLists.txt:
```diff
@@ -168,6 +175,18 @@ foreach( target ${TEST_TARGETS} )
+    else()
+      if( DEFINED ${openmp_LIB_DIR} )
+        set_target_properties( ${target} PROPERTIES BUILD_RPATH "${HIP_CLANG_ROOT}/${openmp_LIB_DIR}" )
+        set_target_properties( ${target} PROPERTIES INSTALL_RPATH "${HIP_CLANG_ROOT}/${openmp_LIB_DIR}" )
```
Would this ever set the binary's install rpath as a relative path, _i.e._, would `${HIP_CLANG_ROOT}` ever be a relative path?

ammallya commented:
Imported to ROCm/rocm-libraries
